File: benchmarks/xlang/src/matmul.py
import time
n = 300
a = [[((i * j) % 100) * 0.01 for j in range(n)] for i in range(n)]
b = [[((i + j) % 100) * 0.01 for j in range(n)] for i in range(n)]
t0 = time.perf_counter()
c = [[0.0] * n for _ in range(n)]
for i in range(n):
    ai = a[i]; ci = c[i]
    for j in range(n):
        s = 0.0
        for k in range(n):
            s += ai[k] * b[k][j]
        ci[j] = s
t = 0.0
for i in range(n):
    for j in range(n):
        t += c[i][j]
print(f"CHECK {round(t)}")
print(f"MS {int((time.perf_counter() - t0) * 1000)}")

File: benchmarks/xlang/src/matmul.R
n <- 300
a <- matrix(0, n, n); b <- matrix(0, n, n); cm <- matrix(0, n, n)
for (i in 1:n) for (j in 1:n) {
  a[i, j] <- (((i - 1) * (j - 1)) %% 100) * 0.01
  b[i, j] <- (((i - 1) + (j - 1)) %% 100) * 0.01
}
t0 <- proc.time()[["elapsed"]]
for (i in 1:n) {
  for (j in 1:n) {
    s <- 0
    for (k in 1:n) s <- s + a[i, k] * b[k, j]
    cm[i, j] <- s
  }
}
t <- 0
for (i in 1:n) for (j in 1:n) t <- t + cm[i, j]
cat("CHECK", format(round(t), scientific = FALSE), "\n")
cat("MS", as.integer((proc.time()[["elapsed"]] - t0) * 1000), "\n")
